pkg/client/introduce/client.go

```diff
@@ -52,77 +60,136 @@ func New(ctx Provider) (*Client, error) {
 		Event:   introduceSvc,
 		service: introduceSvc,
 		store:   store,
+		newUUID: func() string { return uuid.New().String() },
 	}, nil
 }
 
-// SendProposal sends proposal to introducees
+// ProtocolDependency is a helper function that returns the dependency needed for the
+// service to proceed with the protocol. Dependency should be passed to the service through `Continue` function.
+// The function should never return an error, instead of error we provide the callable interface
+// and the state machine will act according to the provided data.
+// Dependency is populated after executing the following functions:
+//  - SendProposal
+//  - SendProposalWithInvitation
+//  - PrepareDependency
+//  - PrepareDependencyWithInvitation
+// usage: e.Continue(c.ProtocolDependency(threadID))
+func (c *Client) ProtocolDependency(thID string) introduce.ActionDependency {
```

Comment:
we haven’t found a way to hide this complexity or rename? i do not think end developers will easily understand this method.

Reply:
The right name for me can be `ActionDependency`. Is it ok?

Reply:
I was avoiding this name because of the function result which clearly says that this is `introduced.ActionDependency`

Reply:
i would rather not expose it at all.

Reply:
At least I need this for the tests (for gomock generation), also it is clear for the user what interface he should implement. There is nothing wrong with exposing interfaces, it is even better.

Reply:
We shouldn't expose service pkg dependencies to the business layer. ie. `introduce.ActionDependency`.

Reply:
@sandrask ^^^

Reply:
@rolsonquadras It is `ActionDependency` and the logic is based on that. This is how the end-user communicates with the protocol. e.g passing the invitation and destinations

Reply:
@soluchok my comments were not on the use of this function. I was merely pointing out that the return type `introduce.ActionDependency` is coming from service package. Till now we have designed in such a way that the consumers will not have any imports from service pkg.

Reply:
Currently, the consumer may impact on the protocol logic. According to that, I must provide the right interface for that with the description of how it works. This is a way how they communicate (end-user with the service) and I cannot just hide it.
What if I need one more function in my dependency? In that case, consumer, client whatever must implement the interface. In my case service requires this interface, the protocol will not work without it. But how can we implement a hidden interface and make sure we satisfy it?
And this is exactly why it is public.

Reply:
To be honest I am not excited about it as well. I would rather do something like that:
* `ProtocolDependency(thID string) Options`
* `Continue(introduce.ActionDependency)`
But the problem is I cannot specify the specific interface for the `Continue` function.

Reply:
@rolsonquadras comment is that you are returning a type from the service package. I would also say that you should prefer to return a struct rather than an interface.

This function could return an exported *Options struct.

Reply:
Done!